d2graph/d2graph.go

```diff
@@ -1224,28 +1219,49 @@ func (g *Graph) SetDimensions(mtexts []*d2target.MText, ruler *textmeasure.Ruler
 			return err
 		}
 
-		obj.Width = float64(go2.Max(defaultDims.Width, desiredWidth))
-		obj.Height = float64(go2.Max(defaultDims.Height, desiredHeight))
-
-		paddingX, paddingY := obj.GetPadding()
+		shapeType := d2target.DSL_SHAPE_TO_SHAPE_TYPE[dslShape]
+		contentBox := geo.NewBox(geo.NewPoint(0, 0), float64(defaultDims.Width), float64(defaultDims.Height))
+		s := shape.NewShape(shapeType, contentBox)
 
+		paddingX, paddingY := s.GetDefaultPadding()
+		// give shapes with icons extra padding to fit their label
+		if obj.Attributes.Icon != nil {
+			labelHeight := float64(obj.LabelDimensions.Height)
+			// Evenly pad enough to fit label above icon
+			paddingX += labelHeight
+			paddingY += labelHeight
+		}
 		switch shapeType {
-		case d2target.ShapeSquare, d2target.ShapeCircle:
-			if desiredWidth != 0 || desiredHeight != 0 {
-				paddingX = 0.
-				paddingY = 0.
+		case shape.TABLE_TYPE, shape.CLASS_TYPE, shape.CODE_TYPE, shape.IMAGE_TYPE:
+		default:
+			if obj.Attributes.Link != "" {
+				paddingX += 32
+			}
+			if obj.Attributes.Tooltip != "" {
+				paddingX += 32
 			}
+		}
 
-			sideLength := math.Max(obj.Width+paddingX, obj.Height+paddingY)
+		// width and height will be set to desired values unless it is too small to fit the content with 0 padding
+		// in which case that minimum size will be used
+		if desiredWidth != 0 || desiredHeight != 0 {
+			paddingX = 0.
+			paddingY = 0.
+			if dslShape != d2target.ShapeText && obj.Attributes.Label.Value != "" {
+				contentBox.Width -= float64(INNER_LABEL_PADDING)
+				contentBox.Height -= float64(INNER_LABEL_PADDING)
+			}
+		}
+		fitWidth, fitHeight := s.GetDimensionsToFit(contentBox.Width, contentBox.Height, paddingX, paddingY)
+		obj.Width = math.Max(float64(desiredWidth), fitWidth)
```

Comment:
i think this fixes https://github.com/terrastruct/d2/issues/699 . can u add to changelog

Reply:
I think this is still an issue, getDefaultSize for image is 128x128 which sets content box dimensions which determines this fit width, so if desired is less than that it still will max to 128

Reply:
fixed with refactor and added to changelog